src/plugins/experimental-discord/createGraph.js

```diff
@@ -187,6 +187,94 @@ function mentionsEdge(message: Model.Message, member: Model.GuildMember): Edge {
   };
 }
 
+function sharedReactionWeights({
+  guild,
+  emojiWeights,
+  roleWeightConfig,
+  wg,
+  memberMap,
+  message,
+  reactions,
+  useAsymptoticReactionWeights,
+}) {
+  let hasEdges = false;
+
+  for (const [ind, reaction] of reactions.entries()) {
+    const emojiRef = Model.emojiToRef(reaction.emoji);
+    let reactionWeight = NullUtil.orElse(emojiWeights[emojiRef], 1);
+    if (useAsymptoticReactionWeights) {
+      // drop off reactionWeight by powers of 2
+      reactionWeight *= 0.5 ** ind;
+    }
+
+    const reactingMember = memberMap.get(reaction.authorId);
+    if (!reactingMember) {
+      // Probably this user left the server.
+      continue;
+    }
+
+    // get the weight of the highest weight role the reacting user has
+    let roleWeight = roleWeightConfig.defaultWeight;
+    const roleWeights = roleWeightConfig.roleWeights;
+    for (const roleRef of reactingMember.roles) {
+      const matchingWeight = roleWeights[roleRef];
+      if (matchingWeight != null && matchingWeight > roleWeight) {
+        roleWeight = matchingWeight;
+      }
+    }
+
+    const node = reactionNode(reaction, message.timestampMs, guild);
+    wg.weights.nodeWeights.set(node.address, roleWeight * reactionWeight);
+    wg.graph.addNode(node);
+    wg.graph.addNode(memberNode(reactingMember));
+    wg.graph.addEdge(reactsToEdge(reaction, message));
+    wg.graph.addEdge(
+      addsReactionEdge(reaction, reactingMember, message.timestampMs)
+    );
+    hasEdges = true;
+  }
+  return hasEdges;
+}
+
+function asymptoticReactionWeights({
+  guild,
+  emojiWeights,
+  roleWeightConfig,
+  wg,
+  memberMap,
+  message,
+  reactions,
+  useAsymptoticReactionWeights,
+}) {
+  // sort reactions from highest to lowest weight emojis
+  const sortedReactions = [...reactions].sort((a, b) => {
+    const bEmojiWeight = emojiWeights[Model.emojiToRef(b.emoji)] || 1;
+    const aEmojiWeight = emojiWeights[Model.emojiToRef(a.emoji)] || 1;
+    return bEmojiWeight - aEmojiWeight;
+  });
+
+  // separate reactions into groups per member (retains sorted order)
+  const mapMemberIdToReactions = new Map();
```

Comment:
To me, this is named more like function rather than a data structure. not a huge deal, just an observation.